Psychtoolbox/PsychHardware/EyelinkToolbox/EyelinkBasic/PsychEyelinkDispatchCallback.m

```diff
@@ -129,326 +161,419 @@
-persistent lasttitle;
-global dh dw offscreen;
-try
-    if ~isempty(eyelinktex)
-        eyerect=Screen('Rect', eyelinktex);
-        % we could cash some of the below values....
-        wrect=Screen('Rect', eyewin);
-        [width, heigth]=Screen('WindowSize', eyewin);
-        dw=round(el.eyeimgsize/100*width);
-        dh=round(dw * eyerect(4)/eyerect(3));
+    function EyelinkDrawClearScreen(eyewin, el)
+        if el.winInfo.StereoMode ~= 0 
+            drawScreens = 2; % stereoscopic drawing
+        else
+            drawScreens = 1; % non-stereoscopic drawing
+        end
+        for it = 0:drawScreens-1 
+            Screen('SelectStereoDrawBuffer', eyewin, it); % select left eye window
+            Screen('FillRect', eyewin, el.backgroundcolour);
+        end
+    end
 
-        drect=[ 0 0 dw dh ];
-        drect=CenterRect(drect, wrect);
-        Screen('DrawTexture', eyewin, eyelinktex, [], drect);
-        %     fprintf('EyelinkDrawCameraImage:DrawTexture \n');
+    function EyelinkDrawInstructions(eyewin, el,msg)
+        if el.winInfo.StereoMode ~= 0 
+            drawScreens = 2; % stereoscopic drawing
+        else
+            drawScreens = 1; % non-stereoscopic drawing
+        end
+        
+        for it = 0:drawScreens-1 
+            Screen('SelectStereoDrawBuffer', eyewin, it); % select left eye window
+            oldFont=Screen(eyewin,'TextFont',el.msgfont);
+            oldFontSize=Screen(eyewin,'TextSize',el.msgfontsize);
+            DrawFormattedText(eyewin, el.helptext, 20, 20, el.msgfontcolour, [], [], [], 1);
+            
+            if el.displayCalResults && ~isempty(msg)
+                DrawFormattedText(eyewin, msg, 20, 150, el.msgfontcolour, [], [], [], 1);
+            end
+            
+            Screen(eyewin,'TextFont',oldFont);
+            Screen(eyewin,'TextSize',oldFontSize);
+        end
+        
+        
     end
-    % imgtitle
-    % if title is provided, we also draw title
-    if ~isempty(eyelinktex) && exist( 'imgtitle', 'var') && ~isempty(imgtitle)
-        %oldFont=Screen(eyewin,'TextFont',el.imgtitlefont);
-        %oldFontSize=Screen('TextSize',eyewin,el.imgtitlefontsize);
-        rect=Screen('TextBounds', eyewin, imgtitle );
-        [w2, h2]=RectSize(rect);
-
-        % added by NJ as a quick way to prevent over drawing and to clear text
-        if newImage || isempty(lasttitle) || ~strcmp(imgtitle,lasttitle)
-            if -1 == Screen('WindowKind', offscreen)
-                Screen('Close', offscreen);
+
+
+
+    function  imgtitle=EyelinkDrawCameraImage(eyewin, el, eyelinktex, imgtitle, newImage)
+        persistent lasttitle;
+        %         global dh dw offscreen;
+        if el.winInfo.StereoMode ~= 0 
+            drawScreens = 2; % stereoscopic drawing
+        else
+            drawScreens = 1; % non-stereoscopic drawing
+        end
+        
+        for it = 0:drawScreens-1 
+            try
+                if ~isempty(eyelinktex)
```

Comment:
All this code inside the for loop only makes sense ``if ~isempty(eyelinktex)``.
So you can skip the whole function a la

```
    if isempty(eyelinktex)
        return;
    end
```

at the very beginning of the function. Even better, EyelinkDrawCameraImage() is not ever called if eyelinktex is empty, so you can save all these isempty checks inside the function and just trust that eyelinktex will be non-empty and there is work to do.

Also, you can move all the calculations out of the for-loop as they apply the same to the left and right and mono images.

Only one `` Screen('SelectStereoDrawBuffer', eyewin, it);`` call at the beginning of the for-loop is needed, not three like now.

Reply:
resolved in 9e17d4f50104aab97855289e617b8627d7442617